benchmark/C/Savina/src/parallelism/Apsp.lf

```diff
@@ -0,0 +1,337 @@
+          cog.outl(f"threads: {threads},")
+      else:
+          cog.outl("threads: 0,")
+    ]]] */
+    threads: 0,
+    /// [[[end]]]
+    build-type : RelWithDebInfo,
+    cmake-include: ["../lib/matrix.cmake"],
+    files: ["../include/matrix.h", "../lib/matrix.c", "../include/PseudoRandom.h"]
+};
+
+preamble {=
+    #include "matrix.h"
+    #include "PseudoRandom.h"
+    #include <math.h>
+=}
+
+reactor ApspFloydWarshallBlock(
+    bank_index: int(0),
+    row_index: size_t(0),
+    graph_size: size_t(300),
+    block_size: size_t(50),
+    dimension: size_t(6),
+    verbose: bool(false))
+{
+    preamble {=
+        
+
+        // Extract (copy) the data block that is relevant for this instance from the given matrix 
+        matrix_t extract_data_from_matrix(matrix_t matrix, size_t block_size, size_t row_offset, size_t col_offset) {
+            matrix_t local_matrix = mat_new_i(block_size, block_size);
+            
+            // FIXME memcpy should be much faster here, but on the other hand this is only done during initialization
+            for(size_t i = 0; i < block_size; i++) {
+                for(size_t j = 0; j < block_size; j++) {
+                    int* value = mat_at_i(matrix, row_offset + i, col_offset + j);
+                    mat_set_i(local_matrix, i, j, *value);
+                }
+            }
+            return local_matrix;
+        }
+    =}
+    
+    state num_neighbors:size_t;
+    state row_offset:size_t; // row offset of the block of this reactor
+    state col_offset:size_t; // column offset of the block of this reactor
+    
+    input start:matrix_t;
+
+    state k:size_t(0); // iteration counter
+    state reported_finish:bool(false);
+
+    input[dimension] from_row:matrix_t;
+    input[dimension] from_col:matrix_t;
+
+    output to_neighbors:matrix_t;    
+    output finished:bool;
+    
+    logical action notify_neighbors:matrix_t;
+    
+    reaction(start) -> notify_neighbors {=
+        // Initialize state variables
+        self->num_neighbors = 2 * (self->dimension - 1);
+        self->row_offset = self->row_index * self->block_size;
+        self->col_offset = self->bank_index * self->block_size;
+        
+        // reset local state
+        self->k = 0;
+        self->reported_finish = false;
+        
+        // start execution
+        matrix_t matrix = extract_data_from_matrix(start->value, self->block_size, self->row_offset, self->col_offset);
+        schedule_copy(notify_neighbors, 0, &matrix, 1);
+    =}
+    
+    reaction(notify_neighbors) -> to_neighbors {=
+        //notify all neighbors
+        SET(to_neighbors, notify_neighbors->value);
+    =}
+    
+    reaction(from_row, from_col) -> notify_neighbors, finished {=
+        // do nothing if complete
+        if(self->k == self->graph_size) {
+            return;
+        }
+        
+        // perform computation
+        matrix_t local_matrix = mat_new_i(self->block_size, self->block_size);
```

Comment:
It doesn't seem like the matrices created during the execution are freed up anywhere.